[Team - 06][BE] 에어비앤비 프로젝트 1주차 2회차 PR

zbqmgldjfh opened the pull request:
## 안녕하세요 Dion! 
우선 저희 팀의 리뷰를 담당해 주셔서 감사의 인사를 전합니다!!

이번 2차 PR을 진행하면서 한가지 고민거리가 있었는데... 이에 대하여 답변해주시면 감사하겠습니다 ㅠ,ㅠ

### 1. Native Query 질문

다음 두 메서드 searchByCondition 과 searchByConditionQueryDsl 은 동일한 일을 수행하고 있습니다.

```java
@Override
public List<House> searchByCondition(Point position, Integer minFee, Integer maxFee) {
  double baseLatitude = position.getX();
  double baseLongitude = position.getY();
  double distance = 1000; // m 단위

  // TODO : *를 필드 명으로 변경해주기
  Query query = em.createNativeQuery("" +
                                     "SELECT * \n" +
                                     "FROM house AS h \n" +
                                     "WHERE (ST_Distance_Sphere(h.point, point(?1, ?2)) < ?3) AND h.price >= ?4 AND h.price <= ?5"
                                     , House.class);

  query.setParameter(1, baseLatitude);
  query.setParameter(2, baseLongitude);
  query.setParameter(3, distance);
  query.setParameter(4, minFee);
  query.setParameter(5, maxFee);

  return query.getResultList();
}
```

다음은 queryDsl 을 통해 조건을 동적으로 받고 싶어 만들어본 함수 입니다.

다만 queryDsl에서 mysql의 함수를 직접 이용하려다 보니 간단하게 사용할수가 없었습니다...

```j
@Override
public List<House> searchByConditionQueryDsl(Point position, Integer minFee, Integer maxFee) {
double baseLatitude = position.getX(); // 내 위치 y
double baseLongitude = position.getY(); // 내 위치 x
double distance = 1000; // m 단위

QHouse subHouse = new QHouse("subHouse");

List<House> results = queryFactory
          .selectFrom(house)
          .where(
                          Expressions.stringTemplate(
                          "function('ST_Distance_Sphere', {0}, {1})", 
                          subHouse.point, GeometryUtils.toPoint(baseLatitude, baseLongitude))
                          .as("subDistance")
                          .loe(String.valueOf(distance)
                      )
           ).fetch();

     return results;
}
```

루시드와 거의 5/26일 하루를 바쳤지만.... queryDsl 상에서 mysql의 함수 사용이 잘 작동하지 않았습니다...

여러 글들을 읽어보니 Expressions.stringTemplate를 사용하라고 하던데.... 인자가 잘 전달되지 않는것 같습니다..

이부분 해결책을 알려주신다면 감사하겠습니다...

    
### 2. nativeQuery에서의 컬럼 명시
native Query를 통해 쿼리 작성 시 전체 컬럼 조회(*) 방식을 지양하고자 했는데, 임베디드 타입의 경우에도 하나 하나 컬럼을 찍어야 하는 불편함이 생겼습니다.

예를 들면 다음과 같습니다.

```
@Id
@GeneratedValue(strategy = GenerationType.IDENTITY)
@Column(name = "house_id")
private Long id;

private String name;
private int price;

@Embedded
private DetailInfo detailInfo;
```

이처럼 Embedded type인 DetailInfo를 가져올 때, DetailInfo 내부에 들어있는 모든 컬럼을 직접 작성해 줘야 합니다. 
반드시 모든 컬럼이 필요한 경우에는 예외적으로 “*”를 작성해도 괜찮을까요?

zbqmgldjfh committed:
[test]: HouseService, HouseServiceTest 구현하기

zbqmgldjfh committed:
[test]: HouseRepository, HouseRepositoryTest 구현하기

zbqmgldjfh committed:
[feat]: 좌표 저장용 Location 클래스 구현

zbqmgldjfh committed:
[feat]: HouseCustomRepository와 impl 구현

- queryDsl과 Native 쿼리 적용을 위한 사용자 repository

zbqmgldjfh committed:
[setting]: 환경 설정 변경

- 지질학 정보를 위한 mysql 엔진 변경과 인포트 추가하기

zbqmgldjfh committed:
[feat]: 각도 저장용 CardinalDirection

zbqmgldjfh committed:
[feat]: 위도, 경도 계산용 GeometryUtils 구현

- 위도, 경도, 각도를 받아 사용자 중심에 일정 거리의 사각형을 구한다.

zbqmgldjfh committed:
[feat]: SearchController 구현하기

- 사용자의 검색 조건을 받아 필터링 하기

zbqmgldjfh committed:
[chore]: House에 Point 삽입하기

- 직접 구현한 Position에서 Point로 변경

leejohy-0223 committed:
[feat]: 연관관계 추가 및 image URL Listing 함수 추가 [#5]

leejohy-0223 committed:
[feat]: 호스팅 역할을 위한 User-House 연관관계 추가 [#5]

leejohy-0223 committed:
[feat]: house 조회 로직 구현 [#5]

leejohy-0223 committed:
[feat]: SQL 함수를 이용하여 범위 검색할 수 있는 native 쿼리 적용

leejohy-0223 committed:
[test]: 범위 검색 test 추가

leejohy-0223 committed:
[refactor]: PR 내용 반영

zbqmgldjfh committed:
[feat]: HouseCustomRepository 와 impl 구현

- HouseCustomRepositoryImpl 의 HouseCustomRepository 구현

zbqmgldjfh committed:
[feat]: Package 위치 변경

- Package Search 에서 House로 변경

zbqmgldjfh committed:
[feat]: GeometryUtils.toPoint 구현

- 위도, 경도를 받아 Point로 변환하는 static 메서드 구현

zbqmgldjfh committed:
[style]: point getter의 이름 변경

zbqmgldjfh committed:
[chore]: @GeneratedValue 전략을 identity 로 명시적으로 지정

zbqmgldjfh committed:
[feat]: 거리 반경 및 가격 조건에 따라 숙소를 검색해주는 쿼리 작성 및 테스트

zbqmgldjfh committed:
[feat]: SearchConditionRequest에 Point 필드를 변경

- 원래 Point 였던 필드를 위도, 경도로 전달하도록 변경

zbqmgldjfh committed:
[feat]: houseRepository 내부의 dto의존성 제거

- houseRepository.findById 에서 원래 dto를 직접 반환하고 있었는데, Entity를 반환하도록 변경

ksundong commented on BE/src/main/java/com/airbnb/api/houses/dto/ImageResponse.java:
```diff
@@ -0,0 +1,17 @@
+
+    public void setUrl(String url) {
+        this.url = url;
+    }
```
`setter` 가 필요한 이유가 있었나요?

zbqmgldjfh replied:
DTO에 setter가 있어도 무방하다 생각하여 사용하게 되었는데,
필요한 시점에만 사용하도록 하겠습니다.

ksundong commented on BE/src/main/java/com/airbnb/api/houses/dto/SearchConditionRequest.java:
```diff
@@ -0,0 +1,78 @@
+    private Double longitude;
+    private Double latitude;
+    private LocalDateTime startDate;
+    private LocalDateTime endDate;
```
타입과 일치하지 않는 네이밍인 이유가 있나요?

zbqmgldjfh replied:
타입과 일치하도록 변수명을 변경하였습니다.

ksundong commented on BE/src/main/java/com/airbnb/api/houses/SearchController.java:
```diff
@@ -0,0 +1,44 @@
+
+@RestController
+@RequestMapping("/api/houses")
+public class SearchController {
```
아직도 `SearchController` 인 이유가 있나요?

zbqmgldjfh replied:
아 함수의 이름까지 변경해야 한다는 인식을 못한것 같습니다. 함수 명도 HouseController 로 변경하였습니다.

ksundong replied:
클래스명을 잘못 말씀하신거겠죵?

ksundong commented on BE/src/main/java/com/airbnb/domain/House.java:
```diff
@@ -30,15 +35,24 @@ public class House {
     @OneToMany(mappedBy = "house")
     private List<WishList> wishLists = new ArrayList<>();
```
여긴 안바뀌었네요. 정말 역참조 할 일이 있나요?

zbqmgldjfh replied:
역참조가 생각해보니 House에서는 필요없는것 같습니다. 사용자만 찜목록에 접근하면 될것 같습니다.
수정하도록 하겠습니다.

ksundong commented on BE/src/main/java/com/airbnb/domain/Location.java:
```diff
@@ -0,0 +1,23 @@
+    private Double longitude;
+
+    public Location() {
+    }
```
이 클래스는 무엇을 위한 클래스인가요?

zbqmgldjfh replied:
더이상 필요없는 도메인이라 제거하도록 하겠습니다.

ksundong commented on BE/src/main/java/com/airbnb/domain/Role.java:
```diff
@@ -1,5 +1,5 @@
 
 public enum Role {
-    USER, GUEST, ADMIN
+    GUEST, HOST, ADMIN
```
이렇게 변경하는건 매우 위험한 선택입니다.
현업에서는 마이그레이션 쿼리와 함께 변경합니다.

zbqmgldjfh replied:
테이블 변경에 대한 주의를 하도록 하겠습니다.

음 근데 그 개발환경에서도 이런 변경은 위험한것 인가요? 서비스 환경에서는 하면 안될것 같은게 느껴지는데, 개발 과정 도중에도 신경을 써야하는지 궁금합니다!!
이게 아직 현업을 못느껴봐서 차이를 못 느끼는점이 큰것 같습니다 ㅠ,ㅠ

ksundong replied:
네 보통 개발하는 환경에서도 배포가 이미 되었다면 배포된 순간부터 신경써야 한다고 생각합니다.

ksundong commented on BE/src/main/java/com/airbnb/domain/User.java:
```diff
@@ -23,4 +23,35 @@ public class User {
     private List<WishList> wishLists = new ArrayList<>();
+
+    @OneToMany(mappedBy = "host")
+    private List<House> hostedHouse;
```
`houses` 가 더 맞지 않을까요?

zbqmgldjfh replied:
아!! 변수명 변경하도록 하겠습니다!!

ksundong commented on BE/src/main/java/com/airbnb/repository/HouseCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,71 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(HouseCustomRepositoryImpl.class);
+
+    private final JPAQueryFactory queryFactory;
+    private final EntityManager em;
```
`EntityManager` 를 직접 사용할 일이 있나요?

zbqmgldjfh replied:
MySQL의 ST_Distance_Sphere 를 직접 사용해보고 싶어 NativeQuery 작성용으로 사용하게 되었습니다.

ksundong replied:
아하 `entityManger`를 사용하고 있었군요. 왜 못봤을까요~~..

ksundong commented on BE/src/main/java/com/airbnb/repository/HouseCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,71 @@
+        double baseLongitude = position.getY();
+        double distance = 1000; // m 단위
+
+        // TODO : *를 필드 명으로 변경해주기
```
네 이건 꼭 해주세요.

zbqmgldjfh replied:
변경 완료 하였습니다!!

ksundong commented on BE/src/main/java/com/airbnb/repository/HouseCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,71 @@
+        double distance = 1000; // m 단위
+
+        // TODO : *를 필드 명으로 변경해주기
+        Query query = em.createNativeQuery("" +
```
굳이 NativeQuery가 필요한가 싶네요. JPA에서 추가 모듈로 공간데이터를 사용하는 기능은 제공할텐데 사용은 해보셨나요?

ksundong replied:
만약에 사용하기 어렵다면, 이 기술이 우리팀에게 정말 최선인지도 고민해보시기 바랍니다.

zbqmgldjfh replied:
MySQL의 ST_Distance_Sphere 를 직접 사용해보고 싶어 사용하게 되었습니다.
호눅스가 수업중에서도 사용해보란 말을 해주셔서 MySQL의 함수를 사용하보려 하고 있습니다.

zbqmgldjfh replied:
해당 기술이 우리팀에 최선인가? 음.... 어려운 질문인것 같습니다... 최선인지를 확인하려면 기준이 필요할탠데,
판단의 기준이라.... 음...

ksundong replied:
네 분명 사용을 위한 학습도 필요하다고 생각하긴 합니다만, 그것이 개발진척을 한 이후에 고민해도 늦지 않겠다는 생각도 드네요.

ksundong replied:
최선인지 판단하는 기준은 "'우리 팀'이 '제한된 기간'동안 '요구사항을 충족'할 수 있는가"가 아닐까요?

ksundong replied:
JPA는 조회성 쿼리에는 약간의 약점이 있어서, 다른 기술을 섞어쓰기도 합니다.
참고가 되셨으면 좋겠네요.

ksundong commented on BE/src/main/java/com/airbnb/service/HouseService.java:
```diff
@@ -18,8 +16,16 @@ public HouseService(HouseRepository houseRepository) {
-    public List<HouseListResponse> findByCondition(SearchConditionRequest request, Pageable pageable) {
-        List<House> houseList = houseRepository.findByCondition(request, pageable);
-        return null;
+    public List<House> findByCondition(Point point, Integer minFee, Integer maxFee) {
```
항상 비즈니스 로직 작성시에 트랜잭션은 필요하진 않은지 생각해보시길 바랍니다.

leejohy-0223 replied:
용도에 맞게 트랜잭션을 적용하도록 하겠습니다.

ksundong commented on BE/src/main/java/com/airbnb/api/houses/SearchController.java:
```diff
@@ -0,0 +1,44 @@
+        House findHouse = houseService.findHouseInformation(id);
+        return new HouseDetailResponse(findHouse);
+    }
+}
```
가급적이면 `Controller` 에 도메인이 등장하지 않도록 해주세요.

leejohy-0223 replied:
서비스계층까지만 도메인을 사용하도록 하겠습니다.

ksundong commented on BE/src/main/java/com/airbnb/utils/GeometryUtils.java:
```diff
@@ -0,0 +1,52 @@
+import org.locationtech.jts.io.ParseException;
+import org.locationtech.jts.io.WKTReader;
+
+public class GeometryUtils {
```
해당 코드에 주석이라도 좀 달아주는게 어떨까요?

ksundong commented on BE/src/main/java/com/airbnb/utils/GeometryUtils.java:
```diff
@@ -0,0 +1,52 @@
+
+public class GeometryUtils {
+
+    public static Location calculateByDirection(Double baseLatitude, Double baseLongitude, double distance,
```
무엇을 계산하기 위한 메서드인가요?

zbqmgldjfh replied:
초기에는 MySQL의 MBRcontains를 사용하려 했기 때문에 만들었는데 더이상 사용하지 않아 제거하였습니다.

ksundong commented on BE/src/main/java/com/airbnb/utils/GeometryUtils.java:
```diff
@@ -0,0 +1,52 @@
+        Double radianLatitude = toRadian(baseLatitude);
+        Double radianLongitude = toRadian(baseLongitude);
+        Double radianAngle = toRadian(bearing);
+        Double distanceRadius = distance / 6371.01;
```
6371.01 은 어떤 수인가요?

zbqmgldjfh replied:
평균 지구 반경에 해당되는 거리 입니다!! 하지만 더이상 사용하지 않아 삭제하였습니다.

https://detegice.github.io/%EA%B8%B0%ED%95%98-%EB%8C%80%EA%B6%8C%EA%B1%B0%EB%A6%AC/

ksundong commented on BE/src/main/java/com/airbnb/utils/GeometryUtils.java:
```diff
@@ -0,0 +1,52 @@
+            return (Point) new WKTReader().read(pointWKT);
+        } catch (ParseException e) {
+            e.printStackTrace();
+        }
```
이런 예외처리는 제발 하지 말아주세요.

leejohy-0223 replied:
새로운 예외 생성 또는 log 출력등의 방식으로 의미있는 예외 처리를 하도록 하겠습니다!

ksundong commented on BE/src/main/resources/application.yml:
```diff
@@ -13,7 +13,8 @@ spring:
         format_sql: true
-    database-platform: org.hibernate.dialect.MySQL8Dialect
+    database: mysql
+    database-platform: org.hibernate.spatial.dialect.mysql.MySQL56InnoDBSpatialDialect
```
왜 5.6 버전에 대응하는 Dialect를 사용하게 되었나요?

leejohy-0223 replied:
참조한 자료들에서 MySQL 공간 데이터 타입을 사용하기 위해 해당 버전을 사용하라고 되어있었습니다! MySQL 8.0 이상에서도 해당 기능을 사용하기 위해서는 5.6버전의 Dialect만을 사용해야 하는 걸로 이해했습니다.

ksundong replied:
8.0에 해당하는 Dialect는 공간데이터를 지원하지 않나용?

leejohy-0223 replied:
다시 찾아보니 org.hibernate.spatial.dialect.mysql.MySQL8SpatialDialect로 변경 가능했습니다! ㅎㅎㅎ 감사합니다.

ksundong commented on BE/src/test/java/com/airbnb/service/HouseServiceTest.java:
```diff
@@ -0,0 +1,63 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest
```
스프링 없이는 테스트할 수 없었나요?

zbqmgldjfh replied:
MySQL의 함수를 끌어다 사용하기 때문에 Slice test가 어려운것 같습니다.

ksundong replied:
`Service`에서 `MySQL`을 알아야한다면, 뭔가 잘못된건 아닐까요?

ksundong commented on BE/src/test/java/com/airbnb/service/HouseServiceTest.java:
```diff
@@ -0,0 +1,63 @@
+        String pointWKT = String.format("POINT(%s %s)", longitude, latitude);
+
+        // WKTReader를 통해 WKT를 실제 타입으로 변환합니다.
+        nowPosition = (Point) new WKTReader().read(pointWKT);
```
그냥 Point 타입을 직접 만들어줄 수는 없었나요?

zbqmgldjfh replied:
음 충분히 적용 가능한 리뷰라 생각합니다만, 이번 프로젝트 에서는 Point를 가져다 사용하기로 하였습니다 ㅠ,ㅠ

ksundong replied:
`Point`는 생성자나 정적 팩토리 메서드가 포함되어있지 않나요?

leejohy-0223 replied:
org.locationtech.jts.geom.Point에는 생성자나 정적 팩토리 메서드가 없어서 위와 같은 방식으로 변환할 수 있다고 참고하였습니다!

ksundong replied:
오잉 그랬군요. 예전에 같은 클래스를 썼던 것 같은데 잘 기억이 안나는군요.

ksundong commented on BE/src/test/java/com/airbnb/service/HouseServiceTest.java:
```diff
@@ -0,0 +1,63 @@
+
+        House house5 = new House("house5", 95000, new DetailInfo(10, "oneRoom5", "방입니다", 4.8, 10), GeometryUtils.toPoint(37.552469, 126.933650));
+        houseService.save(house5); // 1km 보다 먼 거리
+
```
`saveAll` 이라는 것도 있습니다.

leejohy-0223 replied:
service에서 현재 테스트를 위해 saveAll을 만들 필요는 없을 것 같아서, Repository test 시 고려하도록 하겠습니다.

ksundong replied:
기본적으로 제공되는 메서드입니다.

ksundong replied:
무엇을 테스트 해야하는지 특정을 해보시면 좋을 것 같네요.
서비스 영역은 무엇을 표현하기 위한 계층이며, 무엇을 테스트해야하나요?

leejohy-0223 replied:
생각해보니 해당 부분은 Repository에서 작성되어야 할 테스트로 생각됩니다. 
서비스 영역은 비즈니스적 의미를 갖는 로직을 담는다고 우선 생각이 되며, 테스트 시 repository를 Mocking해서 진행해야할 것 같습니다.

ksundong commented on BE/src/test/resources/application.yml:
```diff
@@ -1,18 +1,42 @@
+#  port: 80
+#
+#logging.level:
+#  org.hibernate.SQL: debug
```
왜 안 지우고 주석처리 해주셨나요?

leejohy-0223 replied:
주석 처리는 지우도록 하겠습니다!

ksundong commented on BE/src/main/resources/application.yml:
```diff
@@ -13,7 +13,8 @@ spring:
     properties:
```
jpa의 open-in-view에 대해서도 학습해보시길 바랍니다.

ksundong reviewed (changes requested):
안녕하세요.  Shine & Lucid 리뷰어 dion입니다.

1. `Native Query`가 반드시 필요한 상황인가요?  어쩌면 두분이 쓰시기에 JPA는 아직  배울점이 많은 도구는 아닐까요? 또한 QueryDSL 도입 역시 굳이 필요해보이진 않는 것 같습니다. 제가 단순히 해결책만 알려드린다고 해결이 가능한 것이 아니라, 스스로 트러블 슈팅이 가능하신 툴을 선택해 사용하시는걸 추천드립니다.
2. 네. 예외는 없습니다. 특히 객체에 매핑하는 코드는 반드시 해주세요. 또, Native Query를 작성해야 할 때, 내가 정말 맞는 판단을 하고 있는 건지, 간단한 방법은 정말 없는지, 그래도 안된다면 JPA가 과연 내가 짜려하는 애플리케이션에 적합한 도구인지 생각해보세요.

1주차가 다 끝났네요. 고생하셨습니다.
아쉬운 부분이 곳곳에 눈에 띄어서 변경요청드립니다.
고생많으셨습니다~

zbqmgldjfh committed:
Create main.yml

zbqmgldjfh committed:
[chore]: test 과정 삭제

zbqmgldjfh committed:
[fix]: 오타 수정

zbqmgldjfh committed:
[feat]: S3 설정 추가

zbqmgldjfh committed:
[chore]: GeometryUtils 의 필요없는 메서드 삭제

zbqmgldjfh committed:
[fix] Upload S3 working-directory 수정

zbqmgldjfh committed:
[feat]: CodeDeploy를 위한 appspec.yml 설정 파일 추가

zbqmgldjfh committed:
[feat]: Code Deploy 코드 추가

zbqmgldjfh committed:
[fix] 배포 그룹 명칭 수정

airbnb-code-deploy 으로 수정

zbqmgldjfh committed:
[docs]: PR 수정 todo list 작성

zbqmgldjfh committed:
[fix]: SearchConditionRequest에서 date 변수명 이름 변경

- 기존에는 단순하게 startDate, endDate 였는데 이를 데이터 타입과 일치시켜 startLocalDateTime과 같이 변경하였다.

zbqmgldjfh committed:
[fix]: ImageResponse 에서 setter 제거 하기

zbqmgldjfh committed:
[refactor]: SearchController의 이름을 HouseController 로 변경

zbqmgldjfh committed:
[feat]: House에서 wishList에 대한 역참조 제거하기

zbqmgldjfh committed:
[refactor]: Select 절에 필드 모두 명시해주기

- 기존에는 * 를 사용하고 있었는데, 이를 직접 명시로 모두 변경하였다.

zbqmgldjfh committed:
[chore]: 필요없는 테스트 제거

zbqmgldjfh committed:
[feat]: 필요없는 Location 제거

zbqmgldjfh committed:
[refactor]: hostedHouse에서 houses로 변수명 변경

leejohy-0223 committed:
[refactor]: Service에 Transactional 적용

leejohy-0223 committed:
[refactor]: Service 계층에서만 Domain을 사용하도록 적용

leejohy-0223 committed:
[refactor]: ParseException을 IllegalStateException으로 변경해서 처리

- 예외 처리 핸들러 적용 예정

leejohy-0223 committed:
[refactor]: House에 User(host) 추가 및 테스트 성공하도록 변경

leejohy-0223 committed:
chore: yml에서 사용하지 않는 설정 제거

ksundong commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,60 @@
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: ${{ secrets.AWS_REGION }}
```
👍

ksundong commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,60 @@
+# airbnb-deploy.yml
```
Github Actions를 사용해주셨군용!

ksundong reviewed (approved):
고생하셨습니다. 이번주도 화이팅하세요~~

zbqmgldjfh commented:
@ksundong 
안녕하세요, Dion!!
혹시 그 질문 1번에 대해서 조금 더 알려주실수 있나요? ST_Distance_Spher을 QueryDSL Native Query로 쓰고 싶은데...
아직도 방법을 못찾았습니다..... 조금의 힌트라도 ㅠ,ㅠ
